proxy/proxy/client_test.go

```diff
@@ -122,3 +124,59 @@ func TestConcurrentRefresh(t *testing.T) {
 	}
 	b.Unlock()
 }
+
+func TestMaximumConnectionsCount(t *testing.T) {
+	const maxConnections = 10
+	const numConnections = maxConnections + 1
+	var dials uint64 = 0
+
+	b := &fakeCerts{}
+	certSource := blockingCertSource{
+		map[string]*fakeCerts{}}
+	c := &Client{
+		Certs: &certSource,
+		Dialer: func(string, string) (net.Conn, error) {
+			atomic.AddUint64(&dials, 1)
+
+			// Wait a second to ensure the max connections count is reached by concurrent dialers
+			time.Sleep(time.Second)
```

Comment:
Yup, the sleep is ugly and I was looking for other ways avoiding it too. trying to use a WaitGroup in the Dialer but this was going nowhere...

Your proposal works great and is way simpler, go for it :)